src/main/webapp/app/faq/faq.service.ts

```diff
@@ -0,0 +1,142 @@
+import { Injectable } from '@angular/core';
+import { HttpClient, HttpResponse } from '@angular/common/http';
+import { Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
+import { FAQ, FAQState } from 'app/entities/faq.model';
+import { FAQCategory } from 'app/entities/faq-category.model';
+import { AlertService } from 'app/core/util/alert.service';
+
+type EntityResponseType = HttpResponse<FAQ>;
+type EntityArrayResponseType = HttpResponse<FAQ[]>;
+
+@Injectable({ providedIn: 'root' })
+export class FAQService {
+    public resourceUrl = 'api/courses';
+
+    constructor(
+        protected http: HttpClient,
+        protected alertService: AlertService,
+    ) {}
+
+    create(faq: FAQ): Observable<EntityResponseType> {
+        const copy = FAQService.convertFaqFromClient(faq);
+        faq.faqState = FAQState.ACCEPTED;
+        return this.http.post<FAQ>(`api/faqs`, copy, { observe: 'response' }).pipe(
+            map((res: EntityResponseType) => {
+                return res;
+            }),
+        );
+    }
+
+    update(faq: FAQ): Observable<EntityResponseType> {
+        const copy = FAQService.convertFaqFromClient(faq);
+        return this.http.put<FAQ>(`api/faqs/${faq.id}`, copy, { observe: 'response' }).pipe(
+            map((res: EntityResponseType) => {
+                return res;
+            }),
+        );
+    }
+
+    find(faqId: number): Observable<EntityResponseType> {
+        return this.http.get<FAQ>(`api/faqs/${faqId}`, { observe: 'response' }).pipe(map((res: EntityResponseType) => FAQService.convertFaqCategoriesFromServer(res)));
+    }
+
+    findAllByCourseId(courseId: number): Observable<EntityArrayResponseType> {
+        return this.http
+            .get<FAQ[]>(`${this.resourceUrl}/${courseId}/faqs`, {
+                observe: 'response',
+            })
+            .pipe(map((res: EntityArrayResponseType) => FAQService.convertFaqCategoryArrayFromServer(res)));
+    }
+
+    delete(faqId: number): Observable<HttpResponse<any>> {
+        return this.http.delete<any>(`api/faqs/${faqId}`, { observe: 'response' });
+    }
+
+    findAllCategoriesByCourseId(courseId: number) {
+        return this.http.get<string[]>(`${this.resourceUrl}/${courseId}/faq-categories`, {
+            observe: 'response',
+        });
+    }
+    /**
+     * Converts the faq category json string into FaqCategory objects (if it exists).
+     * @param res the response
+     */
+    static convertFaqCategoriesFromServer<ERT extends EntityResponseType>(res: ERT): ERT {
```

Comment:
Why do we need that? Wouldn't it be possible that the server directly sends using the correct response type/validation?

Reply:
the category is optional, so not every faq response has the field categories